0011-B-Distributed-Lock-building-block.md

```diff
@@ -0,0 +1,359 @@
+  Duration ttl = 3;
+}
+
+message LockResponse {
+  // Lock ID. The app should treat this as an opaque value.
+  string lock_id = 1;
+}
+
+message UnlockRequest {
+  // Resource ID.
+  string resource_id = 1;
+  // Lock ID.
+  string lock_id = 2;
+}
+
+message UnlockResponse {
+  // Empty
+}
+
+message ForceUnlockRequest {
+  // Resource ID.
+  string resource_id = 2;
+}
+
+message ForceUnlockResponse {
+  // Empty
+}
+```
+
+For HTTP (versions are omitted):
+
+```md
+# Lock
+
+**POST /lock/<resource-id>**
+
+Body (JSON):
+
+- `immediate`: bool
+- `ttl`: string (Go duration)
+
+Response (JSON):
+
+- `lockID`: string
+
+# Unlock and Force Unlock
+
+**DELETE /lock/<resource-id>**
+
+Body (JSON):
+
+- `force`: bool
+- `lockID`: string
+
+No response
+```
+
+Pros:
+
+- Probably simpler to implement in the SDKs.
+- Has simple, 1:1 mapping to HTTP methods.
+
+Cons:
+
+- If an app forgets to relinquish a lock, it could be renewed forever by the Dapr sidecar.
+- Apps are not notified if the lock is lost. Doing so would require adding a new method to the app callback channel.
+
+> Note: The fact that apps are not notified if a lock is lost (for example, if renewing a lock fails) may be a significant limitation.  
+> A possible solution could involve adding a new method to the app callback channel to notify the app when a lock is lost. However, this does increase the complexity significantly, as apps are now required to create a separate listener and/or method to handle this scenario. It also requires the app to have an app channel enabled.
+
+#### Option 2: Tied to a long-lived stream
+
+In this case, there are only 2 methods:
+
+- `Lock` is a server-streamed RPC and it's used to acquire a lock:
+  - Apps start by making a call to `Lock`
+  - If `immediate` is false (the default), then the method blocks until a lock is acquired.
+  - If `immediate` is true and the resource is already locked, then an error is returned right away and the stream is closed
+  - After the lock is acquired (regardless of whether it was immediate or not), the caller receives a signal on the RPC.
+  - The RPC remains alive after that.
+  - If the client disconnects, Dapr assumes that the lock is to be relinquished.
+  - If the lock is lost (e.g. after a failed renewal attempt), Dapr sends an error on the RPC and closes it.
+- `ForceUnlock` is invoked to force the removal of a lock.
+
+For gRPC:
+
+```proto
+service Dapr {
+  rpc Lock (LockRequest) returns (stream LockServerStream);
+  rpc ForceUnlock (ForceUnlockRequest) returns (ForceUnlockResponse);
+}
+
+message LockRequest {
+  // When set to true, if the resource is already locked, the method returns right away with an error, and does not wait for the resource to become available.
+  bool immediate = 1;
+  // Resource ID.
+  string resource_id = 2;
+  // If <= 0, uses the default TTL.
+  Duration ttl = 3;
+}
+
+message LockServerStream {
+  oneof lock_server_stream {
+    LockPing ping = 1;
```

Comment:
Why do we need ping? Surely the presence of the stream means that the lock is still active? Or does this mean if the network has failed downstream, worst case scenario the lock is improperly locked for is max ping interval time?

Reply:
The reason for a ping is because it allows the server to send a message over the wire. Without that, if the connection drops, it's not immediately clear to the server that the connection is dropped.